### PR TITLE
Support autocompletion dictionary access in ipython.

### LINKED ISSUE #1628: Add _ipython_key_completions_ to xarray.Dataset
We could use to enable autocompletion for dictionary access like `ds['temp` -> `ds['temperature']`
http://ipython.readthedocs.io/en/stable/config/integrating.html#tab-completion

### PR DESCRIPTION
**Pull request by fujiisoup**
- [x] Closes #1628
 - [x] Tests added / passed
 - [x] Passes ``git diff upstream/master | flake8 --diff``
 - [x] Fully documented, including `whats-new.rst` for all changes and `api.rst` for new API

Support #1628.

### CODE REVIEW

**Comment by fujiisoup** [comment]
It would be really good if we could also have an autocompletion for the keyword argument, such as 
`sel(tem` ->  `sel(temperature=`, but I have no idea though...

**Comment by fujiisoup** [comment]
I think we can also have 
`ds.coords['tem` -> `ds.coords['temperature']` and `ds.data_vars['tem` -> `ds.data_vars['temperature']`.
I will update thsi PR later (maybe few days later).

File: xarray/core/common.py [hunk 1]
```diff
@@ -193,6 +193,11 @@ def __dir__(self):
             if isinstance(item, basestring)]
         return sorted(set(dir(type(self)) + extra_attrs))
 
+    def _ipython_key_completions_(self):
+        item_lists = [item for sublist in self._item_sources for item
+                      in sublist if isinstance(item, basestring)]
```
Comment (shoyer) [inline xarray/core/common.py]: nit: can you reformat this to not line break between "for" and "in", e.g.,
```
        item_lists = [item for sublist in self._item_sources
                      for item in sublist if isinstance(item, basestring)]
```

Actually, additional line breaks might be even better:
```
        item_lists = [item
                      for sublist in self._item_sources
                      for item in sublist
                      if isinstance(item, basestring)]
```

File: xarray/core/dataarray.py [hunk 2]
```diff
@@ -486,6 +486,11 @@ def _attr_sources(self):
         """List of places to look-up items for attribute-style access"""
         return [self.coords, LevelCoordinatesSource(self), self.attrs]
```
Comment (shoyer) [inline xarray/core/dataarray.py]: Should `self.dims` also be in this list? I think so?

File: xarray/tests/test_dataset.py [hunk 3]
```diff
@@ -3505,6 +3505,44 @@ def test_sortby(self):
         actual = ds.sortby(['x', 'y'], ascending=False)
         self.assertDatasetEqual(actual, ds)
 
+    def test_ipython_key_completion(self):
+        ds = create_test_data(seed=1)
+        actual = ds._ipython_key_completions_()
+        expected = ['var1', 'var2', 'var3', 'time', 'dim1',
+                    'dim2', 'dim3', 'numbers']
+        for item in expected:
+            ds[item]
```
Comment (shoyer) [inline xarray/tests/test_dataset.py]: add a comment, e.g.,  `# should not raise`

File: xarray/tests/test_dataset.py [hunk 4]
```diff
@@ -3505,6 +3505,44 @@ def test_sortby(self):
...
+        self.assertTrue(len(actual) == len(expected))
+        self.assertTrue(set(actual) == set(expected))
```
Comment (shoyer) [inline xarray/tests/test_dataset.py]: `assert sorted(actual) == sorted(expected)` would be slightly cleaner than these tests

File: xarray/tests/test_dataset.py [hunk 5]
```diff
@@ -3505,6 +3505,44 @@ def test_sortby(self):
...
+            ds['var1'][item]
+        self.assertTrue(len(actual) == len(expected))
+        self.assertTrue(set(actual) == set(expected))
+
+        actual = ds['var3']._ipython_key_completions_()
+        expected = ['dim3', 'dim1', 'numbers']
+        self.assertTrue(len(actual) == len(expected))
+        self.assertTrue(set(actual) == set(expected))
+
+        # drop variables
```
Comment (shoyer) [inline xarray/tests/test_dataset.py]: I'm not sure this makes sense to test separately.

File: doc/whats-new.rst [hunk 6]
```diff
@@ -74,6 +74,11 @@ Backward Incompatible Changes
 
 Enhancements
 ~~~~~~~~~~~~
+- Support `_ipython_key_completions_` to enable autocompletion for
+  dictionary access.
```
Comment (shoyer) [inline doc/whats-new.rst]: Add an example here?

e.g., `ds['temp` -> `ds['temperature']`

File: xarray/core/dataarray.py (hunk 2 above)
Comment (fujiisoup) [reply xarray/core/dataarray.py]: You are right.
I fixed this and added tests for this.

**Comment by fujiisoup** [comment]
I added a tab-completion support also for `ds.coords` and `ds.data_vars`.

I am getting to feel slight unnaturalness against the following behavior.

```python
In [1]: import xarray as xr
   ...: da = xr.Dataset({'var1': ('x', [0, 1, 2])})

In [2]: 'x' in da
Out[2]: False

In [3]: 'x' in da.keys()
Out[3]: False

In [4]: da['x']
Out[4]: 
<xarray.DataArray 'x' (x: 3)>
array([0, 1, 2])
Dimensions without coordinates: x
```

I understand that it is a designed behavior, 
but it looks a little bit unintuitive.

Maybe `'x' in da` should be `True`?

File: xarray/tests/test_dataset.py [hunk 7]
```diff
@@ -3505,6 +3505,64 @@ def test_sortby(self):
         actual = ds.sortby(['x', 'y'], ascending=False)
         self.assertDatasetEqual(actual, ds)
 
+    def test_attribute_access(self):
+        ds = create_test_data(seed=1)
+        for key in ['var1', 'var2', 'var3', 'time', 'dim1',
+                    'dim2', 'dim3', 'numbers']:
+            self.assertDataArrayEqual(ds[key], getattr(ds, key))
+            assert key in ds.__dir__()
```
Comment (shoyer) [inline xarray/tests/test_dataset.py]: Nit: use `dir(ds)` in preference to `ds.__dir__()`

File: xarray/tests/test_dataset.py [hunk 8]
```diff
@@ -3505,6 +3505,64 @@ def test_sortby(self):
...
+        assert 'foo' in ds['var3'].__dir__()
+
+    def test_ipython_key_completion(self):
+        ds = create_test_data(seed=1)
+        actual = ds._ipython_key_completions_()
+        expected = ['var1', 'var2', 'var3', 'time', 'dim1',
+                    'dim2', 'dim3', 'numbers']
+        for item in actual:
+            ds[item]  # should not raise
+        self.assertTrue(sorted(actual) == sorted(expected))
```
Comment (shoyer) [inline xarray/tests/test_dataset.py]: Just use `assert sorted(actual) == sorted(expected)` (idiomatic when using pytest) or `self.assertEqual(sorted(actual), sorted(expected))`.

**Comment by shoyer** [comment]
@fujiisoup I responded to your remarks on `in` over in https://github.com/pydata/xarray/issues/884#issuecomment-338446509. But yes, I agree.

**Comment by fujiisoup** [comment]
@shoyer 
Thanks. Updated.

**Comment by shoyer** [comment]
Thanks!

**Pull request merged**